File: util/src/test/java/org/cryptomator/util/file/MimeTypesTest.java
package org.cryptomator.util.file;

import org.cryptomator.util.Optional;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.Mockito;

import static org.cryptomator.util.matchers.MimeTypeMatchers.hasMediatype;
import static org.cryptomator.util.matchers.MimeTypeMatchers.hasSubtype;
import static org.cryptomator.util.matchers.OptionalMatchers.anEmptyOptional;
import static org.cryptomator.util.matchers.OptionalMatchers.anOptionalWithValueThat;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.allOf;
import static org.hamcrest.Matchers.is;
import static org.mockito.Mockito.when;

public class MimeTypesTest {

	private static final String AN_EXTENSION = "exe";
	private static final java.lang.String INVALID_MIME_TYPE = "notAMimeTypeBecaseSlashIsMissing";

	private static final java.lang.String MEDIATYPE_OF_VALID_MIME_TYPE = "foo";
	private static final java.lang.String SUBTYPE_OF_VALID_MIME_TYPE = "bar";
	private static final java.lang.String VALID_MIME_TYPE = MEDIATYPE_OF_VALID_MIME_TYPE + '/' + SUBTYPE_OF_VALID_MIME_TYPE;

	private static final String A_FILENAME_WITHOUT_EXTENSION = "fooBarBaz";
	private static final String A_FILENAME_WITH_AN_EXTENSION = "fooBarBaz." + AN_EXTENSION;

	private MimeTypeMap mimeTypeMap = Mockito.mock(MimeTypeMap.class);

	private MimeTypes inTest;

	@BeforeEach
	public void setup() {
		inTest = new MimeTypes(mimeTypeMap);
	}

	@Test
	public void testFromExtensionReturnsEmptyOptionalIfMimeTypeMapReturnsNull() {
		when(mimeTypeMap.getMimeTypeFromExtension(AN_EXTENSION)).thenReturn(null);

		Optional<MimeType> result = inTest.fromExtension(AN_EXTENSION);

		assertThat(result, is(anEmptyOptional()));
	}

	@Test
	public void testFromExtensionReturnsEmptyOptionalIfMimeTypeMapReturnsInvalidMimeType() {
		when(mimeTypeMap.getMimeTypeFromExtension(AN_EXTENSION)).thenReturn(INVALID_MIME_TYPE);

		Optional<MimeType> result = inTest.fromExtension(AN_EXTENSION);

		assertThat(result, is(anEmptyOptional()));
	}

	@Test
	public void testFromExtensionReturnsOptionalContainingCorrectMimeType() {
		when(mimeTypeMap.getMimeTypeFromExtension(AN_EXTENSION)).thenReturn(VALID_MIME_TYPE);

		Optional<MimeType> result = inTest.fromExtension(AN_EXTENSION);

		assertThat(result, is( //
				anOptionalWithValueThat(allOf( //
						hasMediatype(MEDIATYPE_OF_VALID_MIME_TYPE), //
						hasSubtype(SUBTYPE_OF_VALID_MIME_TYPE)))));
	}

	@Test
	public void testFromFilenameReturnsEmptyOptionalIfMimeTypeMapReturnsNull() {
		when(mimeTypeMap.getMimeTypeFromExtension(AN_EXTENSION)).thenReturn(null);

		Optional<MimeType> result = inTest.fromFilename(A_FILENAME_WITH_AN_EXTENSION);

		assertThat(result, is(anEmptyOptional()));
	}

	@Test
	public void testFromFilenameReturnsEmptyOptionalIfMimeTypeMapReturnsInvalidMimeType() {
		when(mimeTypeMap.getMimeTypeFromExtension(AN_EXTENSION)).thenReturn(INVALID_MIME_TYPE);

		Optional<MimeType> result = inTest.fromFilename(A_FILENAME_WITH_AN_EXTENSION);

		assertThat(result, is(anEmptyOptional()));
	}

	@Test
	public void testFromFilenameReturnsEmptyOptionalIfFilenameHasNoExtension() {
		Optional<MimeType> result = inTest.fromFilename(A_FILENAME_WITHOUT_EXTENSION);

		assertThat(result, is(anEmptyOptional()));
	}

	@Test
	public void testFromFilenameReturnsOptionalContainingCorrectMimeType() {
		when(mimeTypeMap.getMimeTypeFromExtension(AN_EXTENSION)).thenReturn(VALID_MIME_TYPE);

		Optional<MimeType> result = inTest.fromFilename(A_FILENAME_WITH_AN_EXTENSION);

		assertThat(result, is( //
				anOptionalWithValueThat(allOf( //
						hasMediatype(MEDIATYPE_OF_VALID_MIME_TYPE), //
						hasSubtype(SUBTYPE_OF_VALID_MIME_TYPE)))));
	}

}

File: util/src/test/java/org/cryptomator/util/file/MimeTypesTest.kt
package org.cryptomator.util.file

import org.cryptomator.util.matchers.MimeTypeMatchers
import org.hamcrest.MatcherAssert
import org.hamcrest.Matchers
import org.junit.jupiter.api.Assertions.assertNull
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.mockito.kotlin.mock
import org.mockito.kotlin.whenever


class MimeTypesTest {

	private val mimeTypeMap: MimeTypeMap = mock()
	private lateinit var inTest: MimeTypes

	@BeforeEach
	fun setup() {
		inTest = MimeTypes(mimeTypeMap)
	}

	@Test
	fun testFromExtensionReturnsEmptyOptionalIfMimeTypeMapReturnsNull() {
		whenever(mimeTypeMap.getMimeTypeFromExtension(AN_EXTENSION)).thenReturn(null)
		val result = inTest.fromExtension(AN_EXTENSION)
		assertNull(result)
	}

	@Test
	fun testFromExtensionReturnsEmptyOptionalIfMimeTypeMapReturnsInvalidMimeType() {
		whenever(mimeTypeMap.getMimeTypeFromExtension(AN_EXTENSION)).thenReturn(INVALID_MIME_TYPE)
		val result = inTest.fromExtension(AN_EXTENSION)
		assertNull(result)
	}

	@Test
	fun testFromExtensionReturnsOptionalContainingCorrectMimeType() {
		whenever(mimeTypeMap.getMimeTypeFromExtension(AN_EXTENSION)).thenReturn(VALID_MIME_TYPE)
		val result = inTest.fromExtension(AN_EXTENSION)
		MatcherAssert.assertThat(
			result, Matchers.`is`( //
				Matchers.allOf( //
					MimeTypeMatchers.hasMediatype(MEDIATYPE_OF_VALID_MIME_TYPE),  //
					MimeTypeMatchers.hasSubtype(SUBTYPE_OF_VALID_MIME_TYPE)
				)
			)
		)
	}

	@Test
	fun testFromFilenameReturnsEmptyOptionalIfMimeTypeMapReturnsNull() {
		whenever(mimeTypeMap.getMimeTypeFromExtension(AN_EXTENSION)).thenReturn(null)
		val result = inTest.fromFilename(A_FILENAME_WITH_AN_EXTENSION)
		assertNull(result)
	}

	@Test
	fun testFromFilenameReturnsEmptyOptionalIfMimeTypeMapReturnsInvalidMimeType() {
		whenever(mimeTypeMap.getMimeTypeFromExtension(AN_EXTENSION)).thenReturn(INVALID_MIME_TYPE)
		val result = inTest.fromFilename(A_FILENAME_WITH_AN_EXTENSION)
		assertNull(result)
	}

	@Test
	fun testFromFilenameReturnsEmptyOptionalIfFilenameHasNoExtension() {
		val result = inTest.fromFilename(A_FILENAME_WITHOUT_EXTENSION)
		assertNull(result)
	}

	@Test
	fun testFromFilenameReturnsOptionalContainingCorrectMimeType() {
		whenever(mimeTypeMap.getMimeTypeFromExtension(AN_EXTENSION)).thenReturn(VALID_MIME_TYPE)
		val result = inTest.fromFilename(A_FILENAME_WITH_AN_EXTENSION)
		MatcherAssert.assertThat(
			result, Matchers.`is`( //
				Matchers.allOf( //
					MimeTypeMatchers.hasMediatype(MEDIATYPE_OF_VALID_MIME_TYPE),  //
					MimeTypeMatchers.hasSubtype(SUBTYPE_OF_VALID_MIME_TYPE)
				)
			)
		)
	}

	companion object {

		private const val AN_EXTENSION = "exe"
		private const val INVALID_MIME_TYPE = "notAMimeTypeBecaseSlashIsMissing"
		private const val MEDIATYPE_OF_VALID_MIME_TYPE = "foo"
		private const val SUBTYPE_OF_VALID_MIME_TYPE = "bar"
		private const val VALID_MIME_TYPE = MEDIATYPE_OF_VALID_MIME_TYPE + '/' + SUBTYPE_OF_VALID_MIME_TYPE
		private const val A_FILENAME_WITHOUT_EXTENSION = "fooBarBaz"
		private const val A_FILENAME_WITH_AN_EXTENSION = "fooBarBaz." + AN_EXTENSION
	}
}
